### PR TITLE
Rework bit_cast

### PR DESCRIPTION
**Pull request by zasdfgbnm**
In the past, `bit_cast` was implemented using `TypelessData` and `erase_type`, which does not always work as observed in https://github.com/NVIDIA/Fuser/pull/627. This PR fixes it by implementing a `std::bit_cast`. Unfortunately, I can not make `std::bit_cast` completely compliant to the C++ standard without using compiler intrinsic, but should be enough for us to use.

Generated kernel for `NVFuserTest.FusionVectorComponentReduce_CUDA`:

```CUDA
__global__ void kernel1(Tensor<std::complex<float>, 1, 1> T0, Tensor<float, 1, 1> T3) {
  NVFUSER_DEFINE_MAGIC_ZERO;
  #pragma unroll 1
  for(nvfuser_index_t i0 = 0; i0 < T0.size[0]; ++i0) {
    Array<float, 2, 1> T1[1];
    T1[0] = {};
    T1[0]
       = std::bit_cast<Array<float, 2, 1>>(T0[(T0.stride[0] * i0)]);
    T3[i0] = 0.00000000000000000e+00;
    #pragma unroll
    for(nvfuser_index_t i1 = 0; i1 < 2; ++i1) {
      float T2[1];
      T2[0] = (T1[0][i1]);
      T3[i0]
        = T3[i0]
        + T2[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
}
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -270,7 +270,7 @@ else()
   # NOTE: our CI build somehow fails with find driver lib.
   target_link_libraries(${NVFUSER_CODEGEN} PUBLIC CUDA::cuda_driver)
 
-  if(UCC_FOUND)  # inherited from pytorch source build
+  if(UCC_FOUND) # inherited from pytorch source build
```
Comment (zasdfgbnm) [inline CMakeLists.txt:273]: Just formatting

File: runtime/basic_type_traits.cu
Comment (zasdfgbnm) [inline runtime/basic_type_traits.cu:1]: Nothing changed. Just formatting.

Comment (zasdfgbnm) [inline runtime/basic_type_traits.cu:1]: Nothing changed, just formatting.

File: runtime/complex_number.cu
Comment (zasdfgbnm) [inline runtime/complex_number.cu:1]: Nothing changed, just formatting.

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**